src/dtype.rs

```diff
@@ -244,9 +246,30 @@ impl_element_scalar!(Complex64 => NPY_CDOUBLE,
 #[cfg(any(target_pointer_width = "32", target_pointer_width = "64"))]
 impl_element_scalar!(usize, isize);
 
-unsafe impl Element for PyObject {
+unsafe impl<T> Element for Py<T>
+where
+    T: PyTypeInfo + 'static,
+{
     const IS_COPY: bool = false;
 
+    fn check_element_types<D: Dimension>(py: Python, array: &PyArray<Self, D>) -> PyResult<()> {
+        // `PyAny` can represent any Python object.
+        if TypeId::of::<PyAny>() == TypeId::of::<T>() {
+            return Ok(());
+        }
+
+        let type_object = T::type_object(py);
+        let iterator = NpySingleIterBuilder::readwrite(array).build()?;
+
+        for element in iterator {
+            if !type_object.is_instance(element)? {
+                return Err(PyDowncastError::new(array, type_name::<T>()).into());
+            }
+        }
```

Comment:
This could be probably written a bit shorter like
```rust
if NpySingleIterBuiler::readwrite(array).build()?.any(|el| !type_obj.is_instance(el)) {
    return Err(...);
}
```

Reply:
Possibly, but I am not sure we can use the iterators at all as they produce references to `T`, but we are still checking whether we actually have `T` or something else entirely.